### PR TITLE
test(unit): update unit test setup, revise existing unit tests

### PR DESCRIPTION
**Pull request by rfprod**
- [x] update jest-preset-angular and respective imports;
- [x] rename setup-jest files - use conventional kebab-case for file names;
- [x] revise existing unit tests - address linting issues, add missing types;

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[x] Maintenance: jest-preset-angular was updated to the next minor version, unit tests were updated accordingly. tslib was updated. Angular integration test dependency versions were fixed.
```

## What is the current behavior?

Unit test setup was updated. `tslib` was updated. Angular integration test dependency versions were fixed. No functional changes have been made.

Issue Number: N/A

## What is the new behavior?

Unit test setup is now a bit more up to date. Some linting issues in the spec files were addressed.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by arturovt** [comment]
@markwhitfeld I'm wondering why CI checks are not running on non-member PRs?

File: packages/store/tests/state-token.spec.ts
Comment (arturovt) [inline packages/store/tests/state-token.spec.ts:18]: Does this need to be async btw?

Comment (rfprod) [reply packages/store/tests/state-token.spec.ts:18]: @arturovt this spec failed sporadically locally on my laptop.
The spec stopped failing sporadically after I made it `async`.
As far as I know all specs that interact with streams (if we want to assert something after stream emits) should be `async`.

The test component has an observable https://github.com/ngxs/store/blob/461a3346e4ec6d2451eb5a7b181270017a0853db/packages/store/tests/state-token.spec.ts#L44

We create the component, then detect changes https://github.com/ngxs/store/blob/461a3346e4ec6d2451eb5a7b181270017a0853db/packages/store/tests/state-token.spec.ts#L55

And expect this to happen https://github.com/ngxs/store/blob/461a3346e4ec6d2451eb5a7b181270017a0853db/packages/store/tests/state-token.spec.ts#L61-L63
^ this failed sporadically without `async` for me locally.

Moreover, these are asynchronous operations obviously https://github.com/ngxs/store/blob/461a3346e4ec6d2451eb5a7b181270017a0853db/packages/store/tests/state-token.spec.ts#L68-L78

I would even suggest to rewrite the above like this
```typescript
store
  .select(TODO_LIST_TOKEN)
  .pipe(take(1))
  .subscribe((value: string[]) => {
      expect(value).toEqual(['hello', 'world']);
  });

store
  .selectOnce(TODO_LIST_TOKEN)
  .subscribe((value: string[]) => {
    expect(value).toEqual(['hello', 'world']);
  });
```

Comment (arturovt) [reply packages/store/tests/state-token.spec.ts:18]: > As far as I know all specs that interact with streams (if we want to assert something after a stream emits) should be async.

> Moreover, these are asynchronous operations obviously

That would've been true if the stream emits asynchronously. The store emits asynchronously only if there's any asynchronous action being handled, otherwise `store.select` always emits synchronously. This is the same as `store.dispatch` with synchronous actions:
```ts
console.log(1);
store.dispatch(new MySyncAction()).subscribe(() => console.log(2));
console.log(3);

// The logs are 1, 2, 3
```

The `async` (`waitForAsync`) is waiting for all micro and macrotasks to be completed within the `AsyncTestZoneSpec`, and there're no scheduled tasks in this test.

If it's the reason to fail, I would re-write it in a `async-await` manner to be more imperative, I personally don't like having subscriptions and expectations within `subscribe` blocks since you never sure if those expects are actually called (unless we do `expect.hasAssertions()`, but it's not reasonable doing that in all tests).

**Comment by rfprod** [comment]
@arturovt does [this change](https://github.com/ngxs/store/pull/1846/commits/9cf11a4ac68a940f1e0920581748b7c46b9f25ac) look good?

**Review by arturovt — APPROVED** [review]
LGTM.

We have to wait until CI checks are run. I don't have any access to our Circle, unfortunately. So for now we `await Mark()` 😂

**Comment by markwhitfeld** [comment]
I will have a look at the CI settings. Don't have my dev machine at the moment, but will check the guthub/circleci settings.

**Comment by rfprod** [comment]
> I will have a look at the CI settings. Don't have my dev machine at the moment, but will check the guthub/circleci settings.

@markwhitfeld what do you think of migrating CI from the Circleci to the GH Actions CI? I can help with that if needed. I have extensive experience with the GH Actions CI.

**Comment by markwhitfeld** [comment]
So frustrated with CircleCI. It seems that something was changed to tighten security at the end of April around PRs originating from forked repositories. But, I have checked and all of the settings are correct. PRs to our repo should be running the build steps! So frustrating.

@rfprod Oh, yes please!!!!!! That would be a massive help. We have been intending to do it for a while.
It would probably be best to start by just mimicking the build tasks we have in CircleCI.
Then we can look at how we can correctly leverage the full power of github actions from there.
Could you create a PR for this?

File: packages/store/tests/state-token.spec.ts
Comment (markwhitfeld) [inline packages/store/tests/state-token.spec.ts:61]: It is preferable for this to either be inside the unit test, or moved out to a `setupFixture` function that returns the specific references that are needed. Was there a specific reason for moving this out of the `it` and using the `beforeEach`?

**Review by markwhitfeld — CHANGES_REQUESTED** [review]

File: packages/store/tests/state-token.spec.ts
Comment (rfprod) [reply packages/store/tests/state-token.spec.ts:61]: @markwhitfeld Of course it can be a matter of preference, there are different ways to do things.
Below are my reasons for making this change.
Each `describe` function is a kind of broader context for several test cases that are wrapped in the `it` functions. In my experience, it's best to remove as many initializations and declarations from the `it` functions as possible to keep the `it` functions concise and easy to read.
Also, the TestBed is initialized and components are compiled asynchronously before each `it` funciton executes. It is a standard and well known pattern to write unit tests as far as I know.
Actually, it is the way the Angular CLI generates empty unit specs. My experience has shown that this approach requires less maintenance costs for large projects.
As for the function that returns needed variables after, I have never used such approach in unit tests. Mainly I use block scoping.

So the main reason is readability and ease of maintenance.

Not that I insist on any specific implementation, you are the maintainer of the project, feel free to make any changes to the PR as necessary.

Comment (rfprod) [reply packages/store/tests/state-token.spec.ts]: @markwhitfeld updated, please check if this works https://github.com/ngxs/store/pull/1846/commits/e09d6c69e86fdace0ac031ee97b48cc32e35d821

**Comment by markwhitfeld** [comment]
@rfprod Could you rebase this on master?

**Comment by rfprod** [comment]
> @rfprod Could you rebase this on master?

@markwhitfeld rebased

File: integrations/hello-world-ng11-ivy-off/src/app/app.component.spec.ts
Comment (arturovt) [inline integrations/hello-world-ng11-ivy-off/src/app/app.component.spec.ts:11]: Since this function is `async` can we use `await` here w/o chaining `then`?

File: integrations/hello-world-ng7/src/app/app.component.spec.ts
Comment (arturovt) [inline integrations/hello-world-ng7/src/app/app.component.spec.ts:18]: Why the `Injector` is provided here?

Comment (rfprod) [reply integrations/hello-world-ng7/src/app/app.component.spec.ts:18]: Good question. This spec was failing without this provider back then when I was refactoring. Idk why it was and way it is not now. Maybe I missed something. I'll remove it in a moment.

**Comment by rfprod** [comment]
> ## BundleMon
> Groups updated (2)
> Status	Path	Size	Limits
> 	@ngxs/store(umd)[gzip]
> ./bundles/*.umd.js
> 43.87KB (+1.4KB **+3.3%**)	+1%
> 	@ngxs/store(umd.min)[gzip]
> ./bundles/*.umd.min.js
> 12.71KB (-1.27KB -9.1%)	+1%
> Unchanged groups (4)
> Final result: x
> 
> [View report in BundleMon website arrow_right](https://app.bundlemon.dev/projects/62b174ff6619780d8caf79fa/reports/62b3269db3092880e18b64cc)
> 
> [Current branch size history](https://app.bundlemon.dev/projects/62b174ff6619780d8caf79fa/reports?branch=update-test-setup&resolution=all) | [Target branch size history](https://app.bundlemon.dev/projects/62b174ff6619780d8caf79fa/reports?branch=master&resolution=days)

@markwhitfeld please take a look. I am not sure how you would prefer to resolve this situation: `umd.js` became larger, exceeding the threshold, `umd.min.js` became smaller.

File: packages/store/tests/state-token.spec.ts
Comment (markwhitfeld) [reply packages/store/tests/state-token.spec.ts]: Thank you for making this update.
Hopefully it also appears as a cleaner approach to you too 😉

File: integrations/hello-world-ng10-ivy/package.json
Comment (rfprod) [reply integrations/hello-world-ng10-ivy/package.json:32]: @markwhitfeld I updated tslib to the latest versions compatible with the corresponding typescript versions in each package.json.

https://github.com/Microsoft/tslib#installing

> For the latest stable version, run:
> ```
> # TypeScript 3.9.2 or later
> yarn add tslib
> 
> # TypeScript 3.8.4 or earlier
> yarn add tslib@^1
>
># TypeScript 2.3.2 or earlier
>yarn add tslib@1.6.1
>```

From the perspective of maintenance and security, as far as I know, it is good to use the latest versions that are compatible.

Comment (rfprod) [reply integrations/hello-world-ng10-ivy/package.json]: @markwhitfeld I reverted dependencies upgrade that I made to make tests work after upgrading jest-preset-angular, now integration tests are failing, in particular this one https://github.com/ngxs/store/runs/7186035903?check_suite_focus=true. Possibly that's the reason why I upgraded tslib in integration tests, I don't remember exactly right now since a lot of time has passed. At least dependencies upgrade that I reverted fixes integration tests.

If this two commits are reverted, integration tests should work again:
- https://github.com/ngxs/store/pull/1846/commits/9ce64c71ed233bc371f28156db1273bfdd84390c
- https://github.com/ngxs/store/pull/1846/commits/7a3ffbc59c00c94fe586d93d0862a41043039e83

Comment (markwhitfeld) [reply integrations/hello-world-ng10-ivy/package.json]: Hmmm... Do you think it could be something to do with the jest preset version that is conflicting with tslib?

Comment (rfprod) [reply integrations/hello-world-ng10-ivy/package.json]: @markwhitfeld most probably.

**Comment by rfprod** [comment]
> Ah, it all makes sense now, the reason why we have been having issues with jest in the integration apps is because the integration apps are borrowing the version from the root node modules folder. It does not make sense to do with jest in integration apps because the version of the jest builder will then be mismatched to the angular version being tested. The integration apps should be using the version of jest that was shipped with that version of angular.
> 
> I guess it is ok for Cypress to still use the shared version because it is pretty isolated from angular code, and has a large install cost.

@markwhitfeld I have several questions, if you will.

How do you want to proceed with dependencies upgrade?
Do you plan to eventually upgrade all project dependencies to the latest versions?
Do you think it is reasonable to support older versions of Angular that are no longer supported by the core Angular team?
https://angular.io/guide/releases#support-policy-and-schedule

<h2 id="support-policy-and-schedule" style="margin-bottom: 0px; font-weight: 500; clear: both; font-size: 3.2rem; margin-top: 4rem; padding-top: 4rem; color: rgb(51, 51, 51); border-top: 1px solid rgb(219, 219, 219); font-family: Roboto, &quot;Helvetica Neue Light&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Lucida Grande&quot;, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Support policy and schedule<a title="Link to this heading" class="header-link" aria-hidden="true" href="https://angular.io/guide/releases#support-policy-and-schedule" style="font-size: inherit; line-height: 3.2rem; font-family: inherit; font-weight: 400; text-decoration: none; color: rgb(110, 110, 110); margin: 0px 6px; user-select: none; visibility: hidden; display: inline-flex; vertical-align: middle;"><i class="material-icons" style="font-family: &quot;Material Icons&quot;; font-weight: normal; font-style: normal; font-size: 24px; line-height: 1; letter-spacing: normal; text-transform: none; display: inline-block; white-space: nowrap; overflow-wrap: normal; direction: ltr; font-feature-settings: &quot;liga&quot;; -webkit-font-smoothing: antialiased;"></i></a></h2><div class="alert is-helpful" style="padding: 16px; margin: 24px 0px; font-size: 1.4rem; width: 800px; box-sizing: border-box; clear: both; color: rgb(68, 68, 68); border-left: 8px solid rgb(25, 118, 210); background-color: rgba(25, 118, 210, 0.05); font-family: Roboto, &quot;Helvetica Neue Light&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Lucida Grande&quot;, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><p style="font-size: 1.6rem; line-height: 3.2rem; font-family: inherit; font-weight: 400; margin: 8px 16px; color: rgb(68, 68, 68);">Dates are offered as general guidance and are subject to change.</p></div><p style="font-size: 16px; line-height: 3.2rem; font-family: Roboto, &quot;Helvetica Neue Light&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Lucida Grande&quot;, sans-serif; font-weight: 400; margin: 1em 0px; color: rgb(68, 68, 68); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">All major releases are typically supported for 18 months.</p>

SUPPORT STAGE | SUPPORT TIMING | DETAILS
-- | -- | --
Active | 6 months | Regularly-scheduled updates and patches are released
Long-term (LTS) | 12 months | Only critical fixes and security patches are released

<p style="font-size: 16px; line-height: 3.2rem; font-family: Roboto, &quot;Helvetica Neue Light&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Lucida Grande&quot;, sans-serif; font-weight: 400; margin: 1em 0px; color: rgb(68, 68, 68); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Angular versions v2 to v11 are no longer under support.</p>Support policy and schedule
Dates are offered as general guidance and are subject to change.

All major releases are typically supported for 18 months.

SUPPORT STAGE | SUPPORT TIMING | DETAILS
-- | -- | --
Active | 6 months | Regularly-scheduled updates and patches are released
Long-term (LTS) | 12 months | Only [critical fixes and security patches](https://angular.io/guide/releases#lts-fixes) are released

The following table provides the status for Angular versions under support.

| VERSION | STATUS | RELEASED | ACTIVE ENDS | LTS ENDS |
| -- | -- | -- | -- | -- |
| ^14.0.0 | Active | 2022-06-02 | 2022-12-02 | 2023-12-02 |
| ^13.0.0 | LTS | 2021-11-04 | 2022-06-02 | 2023-05-04 |
| ^12.0.0 | LTS | 2021-05-12 | 2021-11-12 | 2022-11-12 |

Angular versions v2 to v11 are no longer under support.

**Comment by markwhitfeld** [comment]
@rfprod Thank you so much for walking the long road with this PR. I really appreciate your efforts. And patience!

I know that there is an official support policy for Angular.
While we are not bound to support versions outside of that official policy, it is still good to know what versions we do still support. This is one of the purposes of the integration apps. 
It is not a disaster if an old unsupported version stops working, but at least we know where we stand and can update our official stance on what versions we support.

**Comment by markwhitfeld** [comment]
The ideal would be that we ensure that the jest and tslib deps match the respective version that the respective version of the Angular CLI would have produced. We can find this by running this on our local machine by running something like `npx @angular/cli@11 new ng11` in some folder and then having a look at the `package.json` produced in the `ng11` folder.
For each integrations app we can then use the versions specified for jest as well as the angular dependencies.
I suspect that this would actually resolve the original jest failures that prompted this PR.

PS. In the long run I would actually like to automate this so that we can run a command in the NGXS project like `yarn test:integration 11`, where the `11` is a parameter to the script. This would use the npx command above to generate the project to a target folder using the cli and then copy over the integration app files (from a shared location) and patch the `package.json` file as necessary. It would essentially be an automation of the process above, but using the newly generated folder as the dynamically generated integration app's folder and just the shared integration app files would be in the repo.
Once that is in place we could use the github actions matrix to quickly add whatever versions we desire to the build process.

**Comment by markwhitfeld** [comment]
@rfprod Resurrecting this PR after the upgrade to the new ivy package format.
This resulted in some of the fixes for older versions of integration apps not being needed anymore.
After merging master I found some small tweaks required and made these changes (I pushed to your branch).
If all the checks pass, then I am happy for this to be merged.

**Review by markwhitfeld — APPROVED** [review]

**Pull request merged**
